### DISCUSSION TITLE
Quick question: how to read a new collection in NanoAOD

### DISCUSSION THREAD

**Discussion by gsaha009** [columnflow/columnflow#423] [General]
Hi Columnflow Experts,

I am using a custom made NanoAOD where I need to use some new columns. Along with the `Tau` column, there is `TauProd` columns as well in my ntuples. So, I want to include it in my `events` array. Could you please let me know any possible way to do this?

Thank you

**Comment by pkausw** [comment]
Hi @gsaha009 ,

this is a very interesting problem that will come up more and more in the future I expect. The content of the `events` array and its behavior are constructed using the [coffea NanoAODSchema](https://coffeateam.github.io/coffea/api/coffea.nanoevents.NanoAODSchema.html#nanoaodschema), which also defines which columns are read out. While we have talked about how to potentially implement additional columns/collections based on additional NanoAOD content, I don't think anyone has done/tried that before. Based on a very brief review of the schema, it might be enough to extend the different attributes. However, the technical details is still something that needs to be figured out.

We would be happy to contribute to any developments that you do on your part! However, we would need more information, e.g. a file in the enhanced NanoAOD format that you want to use. Let us know how it goes!

Cheers,
Philip

**Reply by gsaha009** [reply]
Hi @pkausw 
Somehow I managed to access the new column from NanoAOD by adding the name to [this](https://github.com/columnflow/columnflow/blob/master/columnflow/production/cms/seeds.py#L52) list. I am not sure if it is the best way to do it, but I can access the new column after that

**Reply by pkausw** [reply]
Aha. In which context can you access them then? E.g. only in a selector that first calls the `deterministic_seeds` producer, but not outside of the selector? Or globally even in e.g. the `ReduceEvents` or `ProduceColumns` task?

I think it might be the `nTauProd` variable that is created in the for loop you point to in the link above. If it's something like this, then you might want to calculate this once in the `CalibrateEvents` task using a calibrator. But can you use e.g. four momenta logic with this new field? That would surprise me, since this behavior comes from the NanoAODSchema I pointed to above, and that doesn't know anything about a field called `TauProd` 🤔 I'm also not an expert in this matter though

**Comment by pkausw** [comment]
Quick follow up on this @gsaha009 : @nprouvost and I tried accessing NanoAOD files containing a custom column `TauProd` (the ones from the HRareLep CMS group) and had no problems whatsoever accessing any of the information there. So we didn't need to add something to the `deterministic_seed` producer as you pointed out, we simply added the column names to the `uses` set of a selector. Are you sure the behavior you observe doesn't originate from somewhere else?

**Reply by gsaha009** [reply]
Hi @pkausw 
I just tried to run the code without changing anything in `seeds.py`, and it is working. So, you are right. There is no need to change anything in the `columnflow`. Now I just used the new column of NanoAOD in the `uses` dict, and everything is working fine. Sorry for creating the confusion 
Thanks :)

**Reply by pkausw** [reply]
Great, good to hear! Note that you can also use four-momentum logic if you use the `PFCand` behavior of coffea. You can attach this behavior with the `attach_coffea_behavior` producer that comes with columnflow, see the [docs](https://columnflow.readthedocs.io/en/latest/api/production/util.html#columnflow.production.util.attach_coffea_behavior.call_func)
